source/data-modeling/callbacks.txt

```diff
@@ -0,0 +1,173 @@
+reference in the Rails API documentation.
+
+You can implement a callback both top-level and embedded documents. 
+
+.. note:: Callback Invocation Behavior
+
+   For efficiency, {+odm+} invokes the callback only on the document
+   that you performed the persistence action on. This behavior enables
+   {+odm+} to support large hierarchies and handle optimized atomic
+   updates efficiently by not invoking callbacks throughout the document
+   hierarchy.
+
+Take precautions and ensure testability when implementing callbacks for
+domain logic, because these designs can lead to unexpected errors when
+callbacks in the chain halt execution. We recommend using callbacks for
+cross-cutting concerns, such as queueing up background jobs.
+
+Document Callbacks
+------------------
+
+You must implement and register callbacks on your model classes.
+You can register a callback by using ordinary methods, blocks and
+``Proc`` objects, or by defining custom callback objects that use
+classes or modules.
+
+This example demonstrates how to register callbacks on the ``Contact``
+model class in the following ways:
+
+- Includes the ``before_save`` class method which triggers the
+  ``process_phone`` method before a ``Contact`` instance is saved to
+  MongoDB. The ``process_phone`` method is defined separately in the class.
+
+- Includes the ``after_destroy`` class method and uses a block to print a
+  message when a ``Contact`` instance is deleted.
+
+.. literalinclude:: /includes/data-modeling/callbacks.rb
+   :start-after: start-doc-callback
+   :end-before: end-doc-callback
+   :language: ruby
+   :emphasize-lines: 8, 11-13, 16-18
+   :dedent:
+
+The following code performs data operations that demonstrate the
+callback actions:
+
+.. literalinclude:: /includes/data-modeling/callbacks.rb
+   :start-after: start-doc-ops
+   :end-before: end-doc-ops
+   :language: ruby
+   :dedent:
+
+Because callback functionality comes from Active Support, you can use
+the ``set_callback`` class method syntax to register callbacks, as shown
+in the following code:
+
+.. literalinclude:: /includes/data-modeling/callbacks.rb
+   :start-after: start-doc-set-syntax
+   :end-before: end-doc-set-syntax
+   :language: ruby
+   :emphasize-lines: 8-12
+   :dedent:
+
+Association Callbacks
+---------------------
+
+{+odm+} provides the following association callbacks:
+
+- ``after_add``
+- ``after_remove``
+- ``before_add``
+- ``before_remove``
+
+If you register an association callback on your model class, it is
+invoked whenever you add or remove a document from any of the following
+associations:
+
+- ``embeds_many``
+- ``has_many``
+- ``has_and_belongs_to_many``
+
+Specify association callbacks as options on the respective association.
+You must pass the added or removed document as the parameter to the
+specified callback.
+
+The following code demonstrates how to register an association callback
+on a ``User`` model class that embeds multiple ``SavedArticle``
+instances:
```

Comment:
S: might be helpful to describe this code more in-depth, similar to the description for the "Document Callbacks" example. Right now it's not super clear how it works